root/hist/hist/THistFillBenchmarks.cxx

```diff
@@ -16,69 +15,115 @@ static void FillArguments(benchmark::internal::Benchmark *b)
+      hist->Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10));
+   }
+   // Fill with weights
+   for (int i = 0; i < nbOfDataPoints; ++i) {
+      double d = static_cast< double >(i);
+      hist->Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10), d);
+   }
+}
 
 // Benchmark for simple Fill on 2D hist
 static void BM_THist_Fill(benchmark::State &state)
 {
+   int nbOfDataPoints = state.range(0); 
    int nbOfBinsForFirstAxis = state.range(1);
    int nbOfBinsForSecondAxis = state.range(2);
 
-   TH2D hist{"hist", "", nbOfBinsForFirstAxis, 0., 1., nbOfBinsForSecondAxis, 0., 10.};
-
-   int nbOfDataPoints = state.range(0); 
-   for (auto _ : state) {
-      // Fill without weight
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10));
-      }
-
-      // Fill with weights
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10), d);
-      }
-   }
+   for (auto _ : state)
+      FillHist(nbOfDataPoints, histFill[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis]);
 }
 BENCHMARK(BM_THist_Fill) -> Apply(FillArguments);
 
-
-
+// Histograms with different bin configurations for FillN
+TH2D hist11{"hist11", "", 10, 0.1, 1., 10 / 2, 0., 10.};
+TH2D hist12{"hist12", "", 10, 0.1, 1., 100 / 2, 0., 10.};
+TH2D hist13{"hist13", "", 10, 0.1, 1., 1000 / 2, 0., 10.};
+TH2D hist14{"hist14", "", 100, 0.1, 1., 10 / 2, 0., 10.};
+TH2D hist15{"hist15", "", 100, 0.1, 1., 100 / 2, 0., 10.};
+TH2D hist16{"hist16", "", 100, 0.1, 1., 1000 / 2, 0., 10.};
+TH2D hist17{"hist17", "", 1000, 0.1, 1., 10 / 2, 0., 10.};
+TH2D hist18{"hist18", "", 1000, 0.1, 1., 100 / 2, 0., 10.};
+TH2D hist19{"hist19", "", 1000, 0.1, 1., 1000 / 2, 0., 10.};
+
+std::map<int, std::map<int, TH2D*>> histFillN = { { 10, { { 5, &hist11 }, { 50, &hist12 }, { 500, &hist13} } },
+                                    { 100, { { 5, &hist14 }, { 50, &hist15 }, { 500, &hist16 } } },
+                                    { 1000, { { 5, &hist17 }, { 50, &hist18 }, { 500, &hist19 } } } };
+                                    
 // Benchmark for simple FillN on 2D hist
 static void BM_THist_FillN(benchmark::State &state)
 {
+   int nbOfDataPoints = state.range(0); 
    int nbOfBinsForFirstAxis = state.range(1);
    int nbOfBinsForSecondAxis = state.range(2);
 
-   TH2D hist{"hist", "", nbOfBinsForFirstAxis, 0., 1., nbOfBinsForSecondAxis, 0., 10.};
-
-   int nbOfDataPoints = state.range(0); 
-   double * xCoords = new double[nbOfDataPoints];
-   double * yCoords = new double[nbOfDataPoints];
-   double * weights = new double[nbOfDataPoints];
-
-   for (int i = 0; i < nbOfDataPoints; ++i) {
-      double d = static_cast< double >(i);
-      xCoords[i] = d / nbOfDataPoints;
-      yCoords[i] = d / (nbOfDataPoints / 10);
-      weights[i] = d;
-   }
-
    for (auto _ : state) {
```

Comment:
Can you move the map lookup outside the for loop:
```
auto hist = histFillN[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis];
for ...
   hist->FillN
```
That way we don't benchmark the map lookup as much :-)